plugins/toolbox-search/src/toolbox_search.ts

```diff
@@ -0,0 +1,191 @@
+import {BlockSearcher} from './block_searcher';
+
+/* eslint-disable @typescript-eslint/naming-convention */
+
+/**
+ * A toolbox category that provides a search field and displays matching blocks
+ * in its flyout.
+ */
+export class ToolboxSearchCategory extends Blockly.ToolboxCategory {
+  private static readonly START_SEARCH_SHORTCUT = 'startSearch';
+  static readonly SEARCH_CATEGORY_KIND = 'search';
+  private searchField?: HTMLInputElement;
+  private blockSearcher = new BlockSearcher();
+
+  /**
+   * Initializes a ToolboxSearchCategory.
+   * @param categoryDef The information needed to create a category in the
+   *     toolbox.
+   * @param parentToolbox The parent toolbox for the category.
+   * @param opt_parent The parent category or null if the category does not have
+   *     a parent.
+   */
+  constructor(
+      categoryDef: Blockly.utils.toolbox.CategoryInfo,
+      parentToolbox: Blockly.IToolbox,
+      opt_parent?: Blockly.ICollapsibleToolboxItem) {
+    super(categoryDef, parentToolbox, opt_parent);
+    this.initBlockSearcher();
+    this.registerShortcut();
+  }
+
+  /**
+   * Initializes the search field toolbox category.
+   * @returns The <div> that will be displayed in the toolbox.
+   */
+  protected override createDom_(): HTMLDivElement {
+    const dom = super.createDom_();
+    this.searchField = document.createElement('input');
+    this.searchField.type = 'search';
+    this.searchField.placeholder = 'Search';
+    this.workspace_.RTL ? this.searchField.style.marginRight = '8px' :
+        this.searchField.style.marginLeft = '8px';
+    this.searchField.addEventListener('keyup', (event) => {
+      if (event.key === 'Escape') {
+        this.parentToolbox_.clearSelection();
+        return true;
+      }
+
+      this.matchBlocks();
+    });
+    this.rowContents_.replaceChildren(this.searchField);
+    return dom;
+  }
+
+  /**
+   * Returns the numerical position of this category in its parent toolbox.
+   * @returns The zero-based index of this category in its parent toolbox, or -1
+   *    if it cannot be determined, e.g. if this is a nested category.
+   */
+  private getPosition() {
+    const categories = this.workspace_.options.languageTree.contents;
+    for (let i = 0; i < categories.length; i++) {
+      if (categories[i].kind === ToolboxSearchCategory.SEARCH_CATEGORY_KIND) {
+        return i;
+      }
+    }
+
+    return -1;
+  }
+
+  /**
+   * Registers a shortcut for displaying the toolbox search category.
+   */
+  private registerShortcut() {
+    const shortcut = Blockly.ShortcutRegistry.registry.createSerializedKey(
+        Blockly.utils.KeyCodes.B, [Blockly.utils.KeyCodes.CTRL]);
+    Blockly.ShortcutRegistry.registry.register({
+      name: ToolboxSearchCategory.START_SEARCH_SHORTCUT,
+      callback: () => {
+        const position = this.getPosition();
+        if (position < 0) return false;
+        this.parentToolbox_.selectItemByPosition(position);
```

Comment:
Did you have to do this because we don't have a way to select the toolbox category by name?

Reply:
Yes 😔